### PR TITLE
Rich text: formats: allow format to filter value when changing tag name

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Fixes https://github.com/WordPress/gutenberg/pull/34680#issuecomment-934190177.

Problem: when changing the tag name from `span` to `mark` for the highlight format the transparent background color is not added for old highlights that have just a text colour and no background color. The transparent background color is needed to reset the default browser style.

Solution: allow a format type to filter the value to add a transparent background color if none is present.

I'm not quite sure about the future of this API, so I marked it as unstable.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/format-library/src/text-color/index.js
Comment (fullofcaffeine) [inline packages/format-library/src/text-color/index.js:141]: Nit and curious: why the `join( ':' )` and not just a single string?

File: packages/format-library/src/text-color/inline.js
Comment (fullofcaffeine) [inline packages/format-library/src/text-color/inline.js]: Is this `export` really needed? I couldn't find usage of this function outside of this module.

File: packages/rich-text/src/create.js
Comment (fullofcaffeine) [inline packages/rich-text/src/create.js:94]: What are `registered` and `unregistered` attributes in this context?

File: packages/rich-text/src/create.js
Comment (fullofcaffeine) [inline packages/rich-text/src/create.js]: Without this line, `registeredAttributes` would have the following shape in some circumstances:
```
id: undefined
target: undefined
type: undefined
url: "https://wordpress.org/gutenberg"
[[Prototype]]: Object
```

Notice the `id`, `target`, and `type` keys. They're all `undefined`. This caused an exception _(I don't have the backtrace handy at the moment and I don't recall exactly where)_ that would make the inputted character to not to be registered in the editor (hence not part of the block markup). 

Not sure how serious this was, but I guess this could cause more data loss in certain scenarios as shown by the previously failing tests. This is what was causing some tests to fail the snapshot comparison due to differences in the markup (missing a certain char). Let me know what you think about the fix!

File: packages/format-library/src/text-color/index.js
Comment (fullofcaffeine) [inline packages/format-library/src/text-color/index.js]: It seems the logic was inverted here... AFAIU we shouldn't be negating the second expression, right? 🤔

File: packages/rich-text/src/create.js
Comment (fullofcaffeine) [reply packages/rich-text/src/create.js]: The `continue` solution fixed the `undefined` issue, but caused the loop to be skipped in some cases which meant it wouldn't fix the background-color because `_unstableFilterAttributeValue` wouldn't be called. I've changed it to https://github.com/WordPress/gutenberg/pull/35516/commits/e02f6940ae5f4dd5cb70a7532bf88692fadd0af3, which seems to fix the `undefined` value problem depicted above and the original mark background-color issue, too.

File: packages/format-library/src/text-color/index.js
Comment (mcsf) [reply packages/format-library/src/text-color/index.js:141]: It's probably keeping with the style used in `./inline.js`. But personally I'd avoid it in such a situation, since I don't know how smart the VM is.

File: packages/format-library/src/text-color/inline.js
Comment (mcsf) [reply packages/format-library/src/text-color/inline.js]: Maybe a dev artefact? I would remove it too.

**Review by mcsf — APPROVED** [review]
LGTM! Assuming you'll push those additional tests you're working on, feel free to merge.

File: packages/format-library/src/text-color/inline.js
Comment (fullofcaffeine) [reply packages/format-library/src/text-color/inline.js]: Done ✅

**Comment by fullofcaffeine** [comment]
**Do not merge yet!** We found another issue while testing. 

If a post is saved pre-11.7 with a selection with a custom or preset color, it will end up looking as expected in the editor when opened in an instance with this fix applied. **However**, the actual block markup is not updated with the fixes applied at runtime by the logic here. Since it doesn't get persisted, the actual selection will (probably) look wrong in the frontend.

**Review by ockham — CHANGES_REQUESTED** [review]
(see https://github.com/WordPress/gutenberg/pull/35516#issuecomment-945991240)

**Comment by ockham** [comment]
BTW can we add [test fixtures](https://github.com/WordPress/gutenberg/blob/9953bfc6065f395eaa7f8ee43bf3265b53362ef4/test/integration/fixtures/blocks/README.md) for the problematic block markup to cover the [issue](https://github.com/WordPress/gutenberg/pull/34680#issuecomment-934165735) that this is fixing? 🤔

File: packages/rich-text/src/component/index.js
Comment (fullofcaffeine) [inline packages/rich-text/src/component/index.js:171]: This method is a copy of `handleChanges` with the selection logic stripped out. It's possible some of the code below might not be needed, I'll do some further testing, but this seems to work fine for persisting format attributes that are changed upon init, which is the case [here](https://github.com/WordPress/gutenberg/pull/35516/files#diff-6ef6519967765b9d4a49b27938bf474f58270c38d8f5358cf4c878c97e2b4276R137.).

File: packages/rich-text/src/component/index.js
Comment (fullofcaffeine) [inline packages/rich-text/src/component/index.js:193]: It's possible this is not needed at all, didn't have time to test.

**Comment by chad1008** [comment]
This is testing well for me - content added in 11.6 gets highlighted when viewed in the editor on 11.7.0, and isn't properly converted to `mark` tags on the front end if the post is updated.

Once this patch is applied, however, the editor's `mark` tags lose the default yellow highlight as the transparent is applied. Updating the post after this patch is applied successfully converts the `span`s to `mark`s on the front end without losing color and with transparent backgrounds.

I also tested to confirm that any background colors added while on 11.7.0 are not effected by this patch. For both provided presets and custom colors, the transparent background was not applied, and existing backgrounds were respected after the patch was applied.

**Comment by ockham** [comment]
My gut feeling is that some of these changes are happening at a lower level than they should be -- especially the `handleChangesUponInit` happening in the RichText component, and copying most of `handleChange` (which, IIUC, among other things records actions in the undo/redo history). So I'm a bit worried about the ripple effects of such a low-level change.

However, I don't have any evidence to back up that gut feeling. It's really just a feeling that there must be e.g. some sort of allowlist somewhere to which we can add, say, `backgroundColor` and a style attribute that will be persisted.

I hope that another @WordPress/gutenberg-core member with more of a clue about RichText will review this and maybe have some suggestions 😅

**Comment by fullofcaffeine** [comment]
> My gut feeling is that some of these changes are happening at a lower level than they should be -- especially the handleChangesUponInit happening in the RichText component, and copying most of handleChange (which, IIUC, among other things records actions in the undo/redo history). So I'm a bit worried about the ripple effects of such a low-level change.

Thanks for chiming in Bernie! And no problem, I totally understand. I'm also aware this might not be an ideal / permanent solution, but I hope it paves the way to one. Perhaps it could also be used as an interim solution depending on how things go.

> However, I don't have any evidence to back up that gut feeling. It's really just a feeling that there must be e.g. some sort of allowlist somewhere to which we can add, say, backgroundColor and a style attribute that will be persisted.

I'll try to test that assumption!

**Comment by fullofcaffeine** [comment]
> However, I don't have any evidence to back up that gut feeling. It's really just a feeling that there must be e.g. some sort of allowlist somewhere to which we can add, say, backgroundColor and a style attribute that will be persisted.

A quick test doesn't seem to validate the allowlist assumption. I reverted the changes in this changeset and changed from `background-color` to `color` [here](https://github.com/WordPress/gutenberg/pull/35516/files#diff-6ef6519967765b9d4a49b27938bf474f58270c38d8f5358cf4c878c97e2b4276R141) - since `color` is a rule that I know is already applied by existing text-color formats, one could infer that's it's already part of some hypothetical allowlist, if any. The color is applied to the dom, but not synced over to the block markup.

Another aspect of this is that, without the fix here, `background-color` *is* applied to the DOM when added from [here](https://github.com/WordPress/gutenberg/pull/35516/files#diff-6ef6519967765b9d4a49b27938bf474f58270c38d8f5358cf4c878c97e2b4276R141). The only thing missing was "syncing" it over to the block markup, which is now done by the (hackish) solution here. With that in mind, I'd say it's unlikely an allowlist(ish) mechanism exists for format styles (though it's still possible I might be wrong, of course).

**Review by jorgefilipecosta — APPROVED** [review]
Hi @ellatrix, @fullofcaffeine, @mcsf, @ockham, thank you all for the work in fixing this issue!

I pasted created a post with the following markup:
```
<!-- wp:paragraph -->
<p><mark style="background-color:rgba(0, 0, 0, 0)" class="has-inline-color has-purple-color">dfgdf</mark></p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p><mark style="background-color:rgba(0, 0, 0, 0);color:#dd2727" class="has-inline-color">dfdfd</mark></p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p><mark style="background-color:#EEEADD" class="has-inline-color">dfdfd</mark></p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p><mark style="background-color:#EEEADD" class="has-inline-color">dfd</mark></p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>fg<mark style="background-color:#eb0505" class="has-inline-color">f</mark></p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>xzfd <span class="has-inline-color has-green-color">f</span><span class="has-inline-color has-red-color">ghfgh</span></p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p><span style="color:#08a5e9" class="has-inline-color">gfhfg</span></p>
<!-- /wp:paragraph -->
```

The last two paragraphs are old text color markup on trunk they appear on the editor in yellow. In this branch both appear correctly but there is something strange `<p><span style="color:#08a5e9" class="has-inline-color">gfhfg</span></p>`get migrated to a mark markup while `<p>xzfd <span class="has-inline-color has-green-color">f</span><span class="has-inline-color has-red-color">ghfgh</span></p>`keeps the span tag. I think we may have a bug in the migration logic where preset colors are not migrated to marks?

But given that it has no visual impact and the issue is affecting the users, I think we can merge this PR as a temporary solution given that for the users, everything seems to work well.


In the future, I think we should explore other solutions. Namely, when we apply a text color (that may even be applied using a class), we should not add an inline style every time just to make the background transparent. I think we should add the following style on editor and frontend:
```
.has-inline-color:not([class*=background-color]) { background-color:rgba(0, 0, 0, 0) }
```

That rule will avoid adding repeated transparent color inline styles and will make the markup cleaner. We will also remove the need to, during the init phase, apply a background color because the mark is not yellow anymore.

If later we find out we need a programmatic format change on the init, I think a possible higher-level solution going in the direction of what @ockham said would be to on the Format component implementation packages/format-library/src/text-color/index.js, TextColorEdit component have an useEffect call that inspects the value prop. If the useeffect logic finds we need a programmatic change, the useEffect would call onChange with the changes we need.


I found other issues not related to this PR:
- When we apply a preset background color we are not applying the color using a class as happens on the text color. If the preset colors later change, we will have text colors being updated, but the background will not be updated, which may cause design inconsistencies.
- On block colors, when there is a text color, we add has-text-color class. When there is a background color, we add a has-background class. On the format, we add has-inline-color for background or text color. Should we have two distinct classes to be consistent with what happens at the block level?

**Comment by fullofcaffeine** [comment]
This has been cherry-picked into the `release/11.7` branch.

**Comment by fullofcaffeine** [comment]
We're adding tests for this in https://github.com/WordPress/gutenberg/pull/35847.

File: packages/rich-text/src/component/index.js
Comment (ellatrix) [inline packages/rich-text/src/component/index.js:117]: @fullofcaffeine Why is this needed? We shouldn't dirty on init.

Comment (ellatrix) [reply packages/rich-text/src/component/index.js:117]: When we upgrade a span to a mark, it's only visible in the editor. The front-end will still be a span and that's fine. Only when the user changes something in the rich text field will the change be made.

File: packages/rich-text/src/create.js
Comment (ellatrix) [reply packages/rich-text/src/create.js:94]: `unregistered` attributes are attributes that the format has but not registered to it. It basically ensures that, if you e.g. add a style or data attribute to a `strong` tag, it will be preserved.

File: packages/rich-text/src/component/index.js
Comment (fullofcaffeine) [reply packages/rich-text/src/component/index.js:117]: > When we upgrade a span to a mark, it's only visible in the editor. The front-end will still be a span and that's fine. Only when the user changes something in the rich text field will the change be made.

There was a problem with the blocks in the editor not being persisted to the block editor's HTML markup. For instance, the programmatically added `background-color: rgba(0, 0, 0 ,0)` rule wasn't being persisted to the code markup at all even after editing something else/saving the post and this fixed it*. Follow these steps to reproduce:

1. Checkout `v11.6.0` and build;
2. Start a new post, go to the code editor, add two paragraphs: "Custom by 11.6" and "Preset by 11.6". With the "Text color" tool, set a custom color for the first one and a preset for the other. Save and Publish.
3. Checkout `v11.7.0` and build;
4. Open the same post, add two paragraphs below the others: "Custom by 11.7" and "Preset by 11.7". With the "Highlight" tool, set a custom color for the first one and a preset for the other. Save and Publish.
5. Checkout `cbcdfd67f6` (a commit before I added the `handleChangesUponInit`), build;
6. When you load the same post, you'll see something like this:




https://user-images.githubusercontent.com/81248/138486410-e9a91923-429e-45e2-81bc-6bd680f08693.mp4



At first glance, it all looks good in the editor. However, the second paragraph still has a yellow mark on the frontend. By looking at the code markup in the code editor, you'll see that the transparent background color is not there. If you edit the post*, it will still not sync up the markup, it will still look broken in the frontend.

_*It **does** sync up if I edit the **same** paragraph, but that's less than ideal, as it'd require the user to type in each one of them to make the `onChange` to run. That wouldn't scale well in a long post with hundreds or maybe thousands of them._

I'm aware that the `handleChangesUponInit` change is hacky and less than ideal, but it seemed to work well as an interim solution. Feel free to change it as you see fit, I'll also be following up to pick up your brain on it :)

Comment (ellatrix) [reply packages/rich-text/src/component/index.js:117]: That's what it's meant to do. It's not meant to sync up until the user edits the same paragraph. Sure, the old colours that were added will not update to `mark`, but that's fine. All that matters is that they look visually correct in the editor.

Comment (fullofcaffeine) [reply packages/rich-text/src/component/index.js:117]: > That's what it's meant to do. It's not meant to sync up until the user edits the same paragraph. Sure, the old colours that were added will not update to mark, but that's fine. All that matters is that they look visually correct in the editor.

That makes sense! I think the scale of the issue (i.e in number of paragraphs affected) wasn't as big as I thought then. I think it'd be possible to use the solution without syncing upon init, but instruct users to click each affected paragraph, then 🤔 - it'd require more work from users, though. I completely agree that for the long term this shouldn't be here, thanks for creating the follow-up!

**Pull request merged**
